### DISCUSSION TITLE
IEEE 802.11 Quality of Service - Showcase

### DISCUSSION THREAD

**Discussion by adamgeorge309** [inet-framework/inet#619] [Showcases]
Discuss the "IEEE 802.11 Quality of Service" showcase here. Ideas and recommendations are welcome.

**Comment by mgonzalezlopezudc** [comment]
I am wondering if it would be possible to consider a more sophisticated QoS packet classification.

It would be great if the DSCP to UP to AC mappings described in [Diffserv to IEEE 802.11 Mapping](https://tools.ietf.org/id/draft-ietf-tsvwg-ieee-802-11-05.html) could be used in INET.

I have been reviewing the INET source code and I do not see a straigthforward way of doing it. Could you please provide some guidance on this issue?

**Comment by levy** [comment]
In the current master branch, there are two related tags: DscpReq in DscpTag.msg and UserPriorityReq in UserPriorityTag.msg. These tags can be attached to packets in the application directly or using the Socket API or by an intermediate protocol. Similarly, we could add an AccessCategoryReq which would be used by the IEEE 802.11 MAC. The mapping between the requests could be impelemented in separate classifier components. This is partly already implemented in the ExampleQosClassifier, but I admit that the way this currently works is not ideal. It's pretty straightforward to implement the new tag and the required mapping. What do you think?

**Reply by Thats-abhi** [reply]
bro how can i see the implementation code of IEEE 802.11 MAC in inet omnet++, actually i am undergrade student and i have to make a project on this standard, can you please help.

**Comment by mgonzalezlopezudc** [comment]
Thank you very much Levy for your reply. I was not very familiar with tags but now I understand how it works. 

Sounds great what you propose of implementing the new tag and mapping. 

Anyway, I got the feeling that in the Diffserv showcase packet tags are not used to mark packets with DSCP. There, DSCP marking is carried out by means of the DSCPMarker module. Checking the source code in DSCPMarker.cc, DSCP values are set by directly calling method setDscp() from class Ipv4Header (or Ipv6Header). Am I right or am I missing something?

**Comment by levy** [comment]
Yes,  you are right, the DSCPMarker directly modifies the packet. The reason is that the Ipv4 module already added the Ipv4Header to the packet by the time the packet arrives at the DSCPMarker.

**Comment by mgonzalezlopezudc** [comment]
True, you are right.
________________________________

**Comment by ayubshah90** [comment]
Hello Guys!! i am using OMNET 5.6XX version with INET 4.3XX frmakework, i used the same network/application as in this case and tried to re-plot the Graphs. but i can not see the required parameters or vectors details which are needed for throughput, Delay, Qos, jitter etc/. can anyone of you please tell me why is it so? am i missing something or the source code is missing somehting? thanks already

**Reply by Thats-abhi** [reply]
hi brother i dont know the solution of your problem but i am also trying to make a project on QOS just like you, can you please help me how you approached in this project and if you successfully made it please can you guide me. 
I am waiting for your reply
